sdks/python/apache_beam/io/range_trackers_test.py

```diff
@@ -343,6 +344,116 @@ def test_try_split_fails(self):
+    assert key is not None or fraction is not None
+    if fraction is None:
+      fraction = self.key_to_fraction(key, start, end)
+    elif key is None:
+      key = self.fraction_to_key(fraction, start, end)
+
+    if key is None and end is None and fraction == 1:
+      # No way to distinguish from fraction == 0.
+      computed_fraction = 1
+    else:
+      computed_fraction = self.key_to_fraction(key, start, end)
+    computed_key = self.fraction_to_key(fraction, start, end)
+
+    if delta:
+      self.assertAlmostEqual(computed_fraction, fraction,
+                             delta=delta, places=None, msg=str(locals()))
+    else:
+      self.assertEqual(computed_fraction, fraction, str(locals()))
+    self.assertEqual(computed_key, key, str(locals()))
+
+  def test_tiny(self):
+    self._check(fraction=.5**20, key='\0\0\x10')
+    self._check(fraction=.5**20, start='a', end='b', key='a\0\0\x10')
+    self._check(fraction=.5**20, start='a', end='c', key='a\0\0\x20')
+    self._check(fraction=.5**20, start='xy_a', end='xy_c', key='xy_a\0\0\x20')
+    self._check(fraction=.5**20, start='\xFF\xFF\x80',
+                key='\xFF\xFF\x80\x00\x08')
+    self._check(fraction=.5**20 / 3,
+                start='xy_a',
+                end='xy_c',
+                key='xy_a\x00\x00\n\xaa\xaa\xaa\xaa\xaa',
+                delta=1e-15)
+    self._check(fraction=.5**100, key='\0' * 12 + '\x10')
+
+  def test_lots(self):
+    for fraction in (0, 1, .5, .75, 7./512, 1 - 7./4096):
+      self._check(fraction)
+      self._check(fraction, start='\x01')
+      self._check(fraction, end='\xF0')
+      self._check(fraction, start='0x75', end='\x76')
+      self._check(fraction, start='0x75', end='\x77')
+      self._check(fraction, start='0x75', end='\x78')
+      self._check(fraction, start='a' * 100 + '\x80', end='a' * 100 + '\x81')
+    for fraction in (.3, 1/3., 1/math.e, .001, 1e-30, .99, .999999):
+      self._check(fraction, delta=1e-14)
+      self._check(fraction, start='\x01', delta=1e-14)
+      self._check(fraction, end='\xF0', delta=1e-14)
+      self._check(fraction, start='0x75', end='\x76', delta=1e-14)
+      self._check(fraction, start='0x75', end='\x77', delta=1e-14)
+      self._check(fraction, start='0x75', end='\x78', delta=1e-14)
+      self._check(fraction, start='a' * 100 + '\x80', end='a' * 100 + '\x81',
+                  delta=1e-14)
+
+  def test_good_prec(self):
+    self._check(1 / math.e, start='abc_abc', end='abc_xyz',
+                key='abc_i\xe0\xf4\x84\x86\x99\x96',
+                delta=1e-15)
+    self._check(1 / math.e,
+                start='abcd_abc\0\0\0\0\0a',
+                end='abcd_xyz\0\0\0\0\0\0a',
+                key='abcd_i\xe0\xf4\x84\x86\x99\x96',
+                delta=1e-15)
+    self._check(1e-20 / math.e, start='abcd_abc', end='abcd_xyz',
+                key='abcd_abc\x00\x00\x00\x00\x00\x01\x91#\x172N\xbb',
+                delta=1e-35)
+
+
```

Comment:
Is there a more extensive set of tests than https://github.com/apache/incubator-beam/blob/f62d04e22679ee2ac19e3ae37dec487d953d51c1/sdks/java/core/src/test/java/org/apache/beam/sdk/io/range/ByteKeyRangeTrackerTest.java ?